server/controllers/events.js

```diff
@@ -1,61 +1,107 @@
-const Event = require("../models/Event");
+const mongoose = require("mongoose");
+const Joi = require("joi");
+const { Event } = require("../models/Event");
+const httpError = require("../utilities/httpError");
 
 module.exports = {
-  //test function
-  ping: (req, res) => {
-    return res.json({ message: "pong" });
-  },
-  create: async (req, res) => {
+  create: async (req, res, next) => {
     try {
       let data = JSON.parse(req.body.data);
-      data.forEach(obj => obj.user = req.user._id);
+      data.forEach(obj => (obj.user = req.user._id));
       await Event.insertMany(data);
       res.json({ message: "Event created!" });
     } catch (err) {
-      console.log(err);
+      next(err);
     }
   },
-  getAll: async (req, res) => {
+  getAll: async (req, res, next) => {
     try {
-      const events = await Event.find().populate('user').exec();
+      // Get an array of events
+      const events = await Event.find()
+        .populate("user", "displayName")
+        .lean()
+        .exec();
+
       // return all events
       res.json(events);
     } catch (err) {
-      console.log(err);
+      next(err);
     }
   },
-  getOne: async (req, res) => {
+  getOne: async (req, res, next) => {
     try {
-      const event = await Event.findById(req.params.id);
+      const { id } = req.params;
+
+      // Check if the ID is valid
+      if (!mongoose.isValidObjectId(id)) {
+        throw httpError(404);
+      }
+
+      // Get event by id
+      const event = await Event.findById(id).lean().exec();
+
+      // Check if event exists
+      if (!event) {
+        throw httpError(404);
+      }
+
       res.json(event);
     } catch (err) {
-      console.log(err);
+      next(err);
     }
   },
-  deleteEvent: async (req, res) => {
+  deleteEvent: async (req, res, next) => {
     try {
-      const eventId = req.params.id;
-      //checks if an event exists that _id, user, and req.user._id match. This is to prevent users that are authenticated from deleting events they do not author.
-      const event = await Event.findOne({ _id: eventId, user: req.user._id });
-      if (!event) { return res.status(401).send({ message: 'You are not the author of this event' }); }
-      await Event.deleteOne({ _id: eventId });
-      res.json({ message: 'Event deleted' });
-    } catch (error) {
-      console.error(error);
-      res.send(500);
+      const { id } = req.params;
+
+      // Check if the ID is valid
+      if (!mongoose.isValidObjectId(id)) {
+        throw httpError(404);
+      }
+
+      // Prevent users that are authenticated from deleting events they do not author.
+      const event = await Event.findOne({ _id: id, user: req.user._id });
+      if (!event) {
+        throw httpError(401);
```

Comment:
I think [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) has pretty succinct descriptions of each!

Reply:
I guess in this particular case, you want to do two things:
* Checking to see if the event ID exists, and (404 if doesnt exist)
* Check to see if the event is owned by the user (403 if they dont)

You _could_ run two different queries to determine which status code to send back.  In an app where every record (event) isn't publicly accessible, differentiating between the two situations is actually _detrimental_, since if you get a 403, that means that the resource with that ID exists, and then you can try probing for other vulnerabilities by passing that ID in other routes.  For this reason, we generally want to use 404 as a "this doesn't exist _to you_"